### PR TITLE
exporter/prometheusremotewrite: add a Write-Ahead-Log

### PR DESCRIPTION
**Pull request by odeke-em**
Uses an off the shelf WAL implementation to add capabilities
to the Prometheus remote exporter using github.com/tidwall/wal.

By default the WAL will be on and to configure the WAL location,
please use this prometheusremotewrite exporter YAML configuration:

```yaml
exporters:
  prometheusremotewrite:
    endpoint: "http://some.url:9411/api/prom/push"
    wal:
        directory: ./waldir/wal_directory
        truncate_frequency: 200s
        cache_size: 300
```

whose fields are quite similar to Prometheus' WAL fields per https://docs.google.com/document/d/1cCcoFgjDFwU2n823tKuMvrIhzHty4UDyn0IcfUHiyyI/edit#heading=h.mlf37ibqjgov

We are using an off-the-shelf WAL because:
By the time that we get OTLP metrics, we can convert to Prometheus proto,
but trying to implement the Prometheus storage interfaces would involve
either by-passing to-Prometheus-Proto and then save to Prometheus raw Go,
then retrieve Prometheus raw Go and then convert to Prometheus Proto.
It is much easier to get an off the shelf WAL implementation and add
it to the Prometheus implementation.

Fixes https://github.com/open-telemetry/wg-prometheus/issues/9

### CODE REVIEW

**Comment by odeke-em** [comment]
Work in progress

**Comment by bogdandrutu** [comment]
Converting to draft per author comment that this is WIP

File: go.mod
Comment (bogdandrutu) [inline go.mod:40]: Try to avoid unstable (and not very actively maintained) dependencies

**Comment by odeke-em** [comment]
Thanks Bogdan! Just sending up some tests soon and then it’ll be ready.

File: go.mod
Comment (odeke-em) [reply go.mod:40]: AFAIK, this is the most popular general purpose Go WAL implementation and also last change in November 2020, which was 5 months ago, plus I know the author as a reliable person.

**Comment by odeke-em** [comment]
Alrighty, tests added and ready for action!

File: go.mod
Comment (odeke-em) [reply go.mod:40]: The author @tidwall confirmed with me right now that the library is stable and he is a solid maintainer and knows his stuff.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (rakyll) [inline exporter/prometheusremotewriteexporter/exporter.go:33]: Should we be using github.com/prometheus/prometheus/tsdb/wal?

File: exporter/prometheusremotewriteexporter/README.md
Comment (rakyll) [inline exporter/prometheusremotewriteexporter/README.md:46]: Let's try to align with https://docs.google.com/document/d/1cCcoFgjDFwU2n823tKuMvrIhzHty4UDyn0IcfUHiyyI/edit#heading=h.mlf37ibqjgov.

**Review by rakyll — CHANGES_REQUESTED** [review]

**Review by tigrannajaryan — CHANGES_REQUESTED** [review]
There is an initiative to add generic persistent queues to all exporters: https://docs.google.com/document/d/1Y4vNthCGdYI61ezeAzL5dXWgiZ73y9eSjIDitk3zXsU/edit#

I do not think we need to have a custom solution just for this exporter unless we believe it is really necessary due to this exporter having unique requirements (in which case I would like to know what they are).

**Comment by odeke-em** [comment]
>There is an initiative to add generic persistent queues to all exporters: https://docs.google.com/document/d/1Y4vNthCGdYI61ezeAzL5dXWgiZ73y9eSjIDitk3zXsU/edit#
>
>I do not think we need to have a custom solution just for this exporter unless we believe it is really necessary due to this exporter having unique requirements (in which case I would like to know what they are).

Thank you for the comment @tigrannajaryan! We talked about this a few times in a couple of meeting, both the Prometheus Working Group and the Agent working group. Basically we are taking a pragmatic approach to meet our deadlines, AWS want to get the capability in before a deadline that's coming close and whenever the generic WAL is added, the code can simply be moved.

File: exporter/prometheusremotewriteexporter/README.md
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/README.md:46]: All done, thank you!

File: exporter/prometheusremotewriteexporter/wal.go
Comment (tigrannajaryan) [inline exporter/prometheusremotewriteexporter/wal.go:58]: What's the unit for this? Bytes?

File: exporter/prometheusremotewriteexporter/wal.go
Comment (tigrannajaryan) [inline exporter/prometheusremotewriteexporter/wal.go:34]: For uniformness use underscore for word separation in config keys. That's what everything else in the Collector does.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (tigrannajaryan) [inline exporter/prometheusremotewriteexporter/wal.go:32]:
```suggestion
	Dir             string        `mapstructure:"directory"`
```

File: exporter/prometheusremotewriteexporter/README.md
Comment (tigrannajaryan) [inline exporter/prometheusremotewriteexporter/README.md:48]: Explain how to choose a value for this. Alternatively, don't expose it as an option if the user does not have a good way to know how to choose it.

File: exporter/prometheusremotewriteexporter/README.md
Comment (tigrannajaryan) [inline exporter/prometheusremotewriteexporter/README.md:49]: Same here. How to choose it?

File: exporter/prometheusremotewriteexporter/README.md
Comment (tigrannajaryan) [reply exporter/prometheusremotewriteexporter/README.md:46]: Add a comment to explain what "wal" is.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (tigrannajaryan) [inline exporter/prometheusremotewriteexporter/wal.go:76]: Better to use WAL abbreviation as that's what we use everywhere else, including README.
```suggestion
var errWALDisabled = errors.New("WAL is disabled")
```

File: exporter/prometheusremotewriteexporter/wal.go
Comment (tigrannajaryan) [inline exporter/prometheusremotewriteexporter/wal.go:37]: Nit: "create" implies it is something new, always. However, this seems to open an existing wal _or_ creates it. Consider giving the func a better name.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (tigrannajaryan) [inline exporter/prometheusremotewriteexporter/wal.go:113]: Explain in a comment why we need to Sync.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/exporter.go:33]: I believe we talked about this in meetings a few times, the Prometheus WAL needs a Go struct-based prometheus.Registerer which would involve converting from OTLP->Prometheus Go then from Prometheus Go->PrometheusProto which is unnecessary roundtripping.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:146]: >1. See https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#do-not-crash-after-startup

So what do we do? Just log and exit? Sure I can start with that.

>2. Appears the condition is always false so the code is unreachable. Delete the code if so.

Not really AFAIK. When a signal is sent to the program and it causes a shutdown, context.Canceled is emitted, that's the first case; there could be other cases like  a panic, in which case we want to catch it.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:242]: Yeah this was my personal debug

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:58]: Nope, unitless, the count of elements in the cache before truncation.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:34]: So I was asked to make it as uniform with Prometheus as possible, but sure, underscores it is then.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:32]: I had it initially as "dir" but it was requested by @rakyll (and it makes sense to me) to keep it as close to Prometheus as we can per https://docs.google.com/document/d/1cCcoFgjDFwU2n823tKuMvrIhzHty4UDyn0IcfUHiyyI/edit#heading=h.mlf37ibqjgov

File: exporter/prometheusremotewriteexporter/README.md
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/README.md:48]: It's quite the exercise to explain how to do so which is why I provide that link that allows them to learn and choose the value. It is configurable and Prometheus has the same.

File: exporter/prometheusremotewriteexporter/README.md
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/README.md]: Sure, I've updated the comments about what cache_size is. Done, thanks!

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:37]: That's really what it does though, it creates a new WAL every time that it is invoked, and it returns it
Screen Shot 2021-05-05 at 12 51 28 PM

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:58]: It is definitely an odd variable name but mirroring what the actual WAL uses.

Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:58]: I've renamed it to NBeforeTruncation.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go]: Done, thank you!

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go]: It isn't incorrect, a WAL is firstly flushed, aka commit the entries at the end to disk; truncate from the front the entries that have been exported out.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:117]: Sure, done, thanks!

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go]: Removed, done, thanks!

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:213]: Nah the idea is to retry no too frequently but also not too soon; if for example the WAL is locked in use by another process for say 1hr, I think it is definitely alright to keep trying.

>Alternatively, is there a way to wait on the readiness event directly instead of polling?

A bit of overkill for an already complex PR that has taken 2+ weeks to get merged. Perhaps later on but that isn't a core problem and also integrating file watching has its own separate cycle of review. For now this will suffice.

File: exporter/prometheusremotewriteexporter/README.md
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/README.md]: Done, thanks!

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go]: Cool, removed, thank you @tigrannajaryan!

**Comment by odeke-em** [comment]
Reviews addressed, please help me with another wave @rakyll @bogdandrutu @tigrannajaryan and thank you!

File: exporter/prometheusremotewriteexporter/wal.go
Comment (tigrannajaryan) [reply exporter/prometheusremotewriteexporter/wal.go]: Given that this is a temporary solution and will be replaced by generic storage extension (we agree on that, didn't we? if not can't merge :-) ) I suggest that we make this uniform with storage extension configuration option (and it uses `directory).
I believe we should generally aim for uniformness within the Collector, and only then consider the uniformity with a particular 3rd party technology we support.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (tigrannajaryan) [reply exporter/prometheusremotewriteexporter/wal.go]: New comment LGTM.

File: exporter/prometheusremotewriteexporter/README.md
Comment (tigrannajaryan) [reply exporter/prometheusremotewriteexporter/README.md]: You mean this link https://golang.org/pkg/time/#ParseDuration ? That just describes the format of the field. I did not mean to describe that. I mean how does the user know what duration to use? Should I use 1 millisecond? 1 hour? 1 day? Some hints would be useful.

**Review by tigrannajaryan — COMMENTED** [review]
LGTM, can be merged if other reviewers also approve.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (rakyll) [inline exporter/prometheusremotewriteexporter/exporter.go:305]: Can we name this something like exportWriteRequests?

File: exporter/prometheusremotewriteexporter/wal.go
Comment (rakyll) [reply exporter/prometheusremotewriteexporter/wal.go:211]: Maybe, we can add a TODO to log something here as a follow up.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (rakyll) [inline exporter/prometheusremotewriteexporter/wal.go:185]: How are we flushing the WAL at a Shutdown?

**Review by rakyll — APPROVED** [review]
This generally looks good, added a few comments. We can follow up with improvements.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:185]: On shutdown, the WAL's .Close() method is invoked and that toggles a shutdown, before invoking this method in (*PrwExporter).turnOnWALIfEnabled(), we derive a cancellable context.Context and after the prwe.closeChan receives a value, it also invokes cancel() which ensures that this method will terminate alright.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/exporter.go]: Done, thanks!

**Comment by odeke-em** [comment]
@tigrannajaryan @rakyll @bogdandrutu addressed please take a look, thank you!

**Review by tigrannajaryan — APPROVED** [review]
Please rebase from main and make sure the build passes.

**Comment by bogdandrutu** [comment]
@alolita this is not ready to merge, we need to have "required" checks green, and tests are failing in this PR.

**Comment by alolita** [comment]
@bogdandrutu Acknowledged. Will work w @odeke-em to get these tests passing.

**Comment by odeke-em** [comment]
Roger that. Windows is tripping out with t.TempDir() exit, then am working
on sorting the results by Label before comparison.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (pmm-sumo) [inline exporter/prometheusremotewriteexporter/wal.go:126]: It's more like a question than a comment. My understanding is that `retrieveWALIndices` essentially reloads the WAL from disk and resets the read/write indexes accordingly. Why is that necessary? It seems somewhat heavy to me as it opens the file, etc. but I might miss something

Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:126]: Basically there is one end that reads and then one that writes. This routine is invoked after we've flushed out to disk all prior entries. The idea is that afterwards, we want a true definition of where the read and write indexes are in the WAL and not just in RAM. It is needed otherwise the indices in RAM could be stale. This routine isn't invoked often and no it isn't heavy.

**Comment by odeke-em** [comment]
@bogdandrutu ready for review; the failing tests are seemingly with problems that already exist as seen in other PRs (memory ballast tests are flaky) but then there is one where the contrib tests somehow try to use an old module of this repo yet we are adding fresh code.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (pmm-sumo) [reply exporter/prometheusremotewriteexporter/wal.go:126]: Thanks @odeke-em for the explanation! There's just one thing I am unsure of - when the indices in RAM could go stale? Is there a valid scenario when this could happen or is it more like a precaution?

File: exporter/prometheusremotewriteexporter/wal.go
Comment (pmm-sumo) [inline exporter/prometheusremotewriteexporter/wal.go:97]: I think the error message should mention `first WAL index`?

File: exporter/prometheusremotewriteexporter/wal.go
Comment (pmm-sumo) [inline exporter/prometheusremotewriteexporter/wal.go:101]:
```suggestion
		return fmt.Errorf("prometheusremotewriteexporter: failed to retrieve the last WAL index: %w", err)
```

**Comment by bogdandrutu** [comment]
@odeke-em please fix these last comments and I will merge it.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:126]: When truncation of indices from the front occurs, it is up to the discretion of the tidwall code to change indices, hence we need to refresh.

**Comment by odeke-em** [comment]
Thank you @pmm-sumo and @bogdandrutu, addressed and please take a look.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (bogdandrutu) [inline exporter/prometheusremotewriteexporter/exporter.go:316]: Why async? What is the motivation for this change?

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (bogdandrutu) [inline exporter/prometheusremotewriteexporter/exporter.go:99]: This should happen in `Start` of the component not during creation.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (bogdandrutu) [inline exporter/prometheusremotewriteexporter/exporter.go:110]: It is guaranteed by the framework that Shutdown is called only once.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (bogdandrutu) [inline exporter/prometheusremotewriteexporter/exporter.go:309]: It is not a go pattern to return slice of errors.

File: exporter/prometheusremotewriteexporter/config.go
Comment (bogdandrutu) [inline exporter/prometheusremotewriteexporter/config.go:34]: Please explain how this interacts with "sending_queue". During the SIG meeting we talked about this as a replacement for sending queue IIRC

File: exporter/prometheusremotewriteexporter/wal.go
Comment (bogdandrutu) [inline exporter/prometheusremotewriteexporter/wal.go:151]: Why different go-routine? `turnOnWALIfEnabled` should completely finish when this returns correct? Otherwise things like `shutdownCtx` may be already closed.

**Review by bogdandrutu — COMMENTED** [review]
Please consider for easier review to split PRs like this into:
1. Adding the WAL implementation
2. Enabling the WAL to the component.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (alolita) [reply exporter/prometheusremotewriteexporter/wal.go:151]: @odeke-em can you respond to Bogdan's questions? Thx.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/exporter.go:309]: Following precedent with the prior code
Screen Shot 2021-05-18 at 10 39 53 AM


It is also a normal Go idiomatic code to return multiple errors uncombined if you are processing multiple operations. Please see:
* https://pkg.go.dev/cloud.google.com/go/bigtable#Table.ApplyBulk

Later on however, we could use Hashicorp's multierror but for now keeping code as it was before, and just adapting.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/exporter.go:316]: We want to begin processing requests immediately without having to wait to fill up the buffered channel. However, I shall revert it.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:151]: No, readFromWALThenExport will concurrently run throughout the lifetime of the program,
until shutdownCtx is canceled. turnOnWALIfEnabled just triggers the code that'll run until shutdown.

>Otherwise things like shutdownCtx may be already closed.

If it is canceled, readFromWALThenExport will terminate.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/exporter.go:99]: AFAIK there isn't a .Start for this exporter.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/exporter.go:110]: I added this for extra protection because if someone accidentally invokes .Shutdown() in a defer and then later after processing an error, we'll have issues with the WAL and crashes too. This change brings even better protections, so I'd highly recommend we use the .closed guard.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (bogdandrutu) [reply exporter/prometheusremotewriteexporter/wal.go:151]: A component should be entirely shutdown (complete) when the Shutdown returns and should not have any other work scheduled for later. We want every component to be "restartable" and this will make it impossible to do that.

See https://github.com/open-telemetry/opentelemetry-collector/blob/main/component/component.go#L60

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (bogdandrutu) [reply exporter/prometheusremotewriteexporter/exporter.go:95]: Then you should add one :)

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (bogdandrutu) [reply exporter/prometheusremotewriteexporter/exporter.go:106]: Shutdown is not deferred, and needs specific guarantees, see the other comment.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (bogdandrutu) [reply exporter/prometheusremotewriteexporter/exporter.go:310]: Sure, we can address later.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/wal.go:36]: I had it before as CacheSize but it didn't make sense for code and some folks asked for clarification.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (odeke-em) [reply exporter/prometheusremotewriteexporter/exporter.go:106]: @bogdandrutu can people import this code or not? If they can import specific components, they will misuse it. I honestly don't think making it more brittle is reasonable, yet preventing people from shooting their foot, but if that matches the review I'll change it.

File: exporter/prometheusremotewriteexporter/wal.go
Comment (bogdandrutu) [reply exporter/prometheusremotewriteexporter/wal.go:36]: Then the mapstructure name is not good, either one or the other.

**Review by bogdandrutu — CHANGES_REQUESTED** [review]
because it is hard to make progress on the entire PR, I would suggest to split this into 2 PRs:
1. Adds only the WAL implementation.
2. Connects it to the component.

**Comment by rakyll** [comment]
It seems like adding WAL to the existing component is a few lines, the majority of the change is the WAL implementation. @bogdandrutu, are you asking the internals to be structured differently so the WAL stands alone as a generic WAL? I am thinking it might cause new unnecessary abstractions.

**Comment by bogdandrutu** [comment]
I don't have any other way to make the author listening to suggestions... I tried and explain in bunch of places that we don't want to add things like:
* https://github.com/open-telemetry/opentelemetry-collector/pull/3017/files#r633797195
* https://github.com/open-telemetry/opentelemetry-collector/pull/3017/files#r633799397

Please help me understand how else I can review this. I think right now even may seem a little there are probably 100 lines to the component and 400 for wal, so reducing by 20% helps, when I see that I really need to look careful and fight for every change.

**Comment by odeke-em** [comment]
>Please help me understand how else I can review this. I think right now even may seem a little there are probably 100 lines to the component and 400 for wal, so reducing by 20% helps, when I see that I really need to look careful and fight for every change.

Alright, thanks for letting me know your response @bogdandrutu, thanks for chiming in @rakyll. I am going to try again afresh one more time in the next couple of hours; it is 1:15AM here.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (codesome) [reply exporter/prometheusremotewriteexporter/exporter.go:34]: github.com/prometheus/prometheus/tsdb/wal is actually a generic WAL implementation which takes byte slices (similar to tidwall/wal but without the index). It would be interesting (and looking forward!) to see performance of `tidwall/wal` with remote write :)

**Pull request closed without merging**
